routing/turns_tts_text.cpp

```diff
@@ -45,10 +46,195 @@ void GetTtsText::ForTestingSetLocaleWithJson(std::string const & jsonBuffer, std
   m_getCurLang = platform::ForTestingGetTextByIdFactory(jsonBuffer, locale);
 }
 
+// Returns -1 for not found, positive for which item in the haystack was found
+long FindInStrArray(const std::vector<std::string>& haystack, std::string& needle)
+{
+  auto begin = haystack.begin();
+  for(std::vector<std::string>::const_iterator i = begin; i != haystack.end(); ++i) {
+    std::string item = *i;
+    std::string::size_type idx = needle.find(item);
+    if (idx != std::string::npos) {
+      long item_i = std::distance( begin, i );
+      LOG(LINFO, ("TTSf found", item_i, item, idx, needle));
+      return item_i;
+    }
+  }
+  return -1;
+}
+
+void HungarianBaseWordTransform(std::string& myString) {
+  std::string_view baseStr[] = {"e", "a", "ö", "ü"};
+  std::string_view harmonyStr[] = {"é", "á", "ő", "ű"};
+
+  for (size_t i = 0; i < std::size(baseStr); ++i)
+  {
+    if (strings::EndsWith(myString, baseStr[i]))
+      myString.replace(myString.size() - baseStr[i].size(), baseStr[i].size(), harmonyStr[i]);
+  }
+}
+
+uint8_t CategorizeHungarianAcronymsAndNumbers(std::string const & myString)
+{
+  std::vector<std::string> backNames = {
+      "A", // a
+      "Á", // á
+      "H", // há
+      "I", // i
+      "Í", // í
+      "K", // ká
+      "O", // o
+      "Ó", // ó
+      "U", // u
+      "Ű", // ú
+      "0", // nulla or zéró
+      "3", // három
+      "6", // hat
+      "8", // nyolc
+  };
+
+  std::vector<std::string> frontNames = {
+      // all other letters besides H and K
+      "B", "C", "D", "E", "É", "F", "G", "J", "L", "M", "N", "Ö", "Ő", "P", "Q", "R", "S", "T", "Ú", "Ü", "V", "W", "X", "Y", "Z",
+      "1", // egy
+      "2", // kettő
+      "4", // négy
+      "5", // öt
+      "7", // hét
+      "9", // kilenc
+  };
+
+  std::vector<std::string> specialCaseFront = {
+      "10", // tíz special case front
+      "40", // negyven front
+      "50", // ötven front
+      "70", // hetven front
+      "90", // kilencven front
+  };
+
+  std::vector<std::string> specialCaseBack = {
+      "20", // húsz back
+      "30", // harminc back
+      "60", // hatvan back
+      "80", // nyolcvan back
+  };
+
+  //'100', // száz back
+
+  strings::UniString original = strings::MakeUniString(myString);
+  for (std::string::size_type i = myString.size()-1; i > 0; i--) {
+    // special is 2 char, so check last 2
+    std::string oneBuf = strings::ToUtf8(strings::Substring(original, i, 1));
```

Comment:
What is going on here? Looks like an unnecessary overhead. Can you explain the algorithm?

Reply:
@biodranik if the last two characters in an acronym or number (i.e. we won't say ABC or 123 as if they were words, we will spell it out like ay bee see or one hundred twenty three) then in Hungarian we start by looking at the last two characters. If the last two characters are 10, 40, 50, 70, 90 then we have a "-re" ending because of how it's pronounced. If they're 20, 30, 60, 80 then they'll have a "-ra" ending.

A phrase ending in "-hundred" is a special case, so if the last three letters are "100" then that has a "-ra" ending.

If none of the above are true, then we can simply look at the last character in the string for the appropriate suffix. If the last character is one of AÁHIÍKOÓUŰ0368 then it gets a "-re" ending. All other cases will get a "-ra" ending however we can't simply stop there because if there is some unknown character like whitespace or punctuation we keep going until we find a match or we run off the end of the word (`" "`) which is a suspicious condition (what Hungarian word or acronym doesn't have a single character in the Hungarian alphabet?)

I suppose it's possible to make this with some pretty crazy regex, just remember this isn't about finding any occurrence of a string or the last occurrence or even the last character, this is about starting at the end of a word or acronym and *counting backwards* until we find a matching *ending word-sound* so that we know whether 123 or 110 or 100 or 1000 or ABC or ABA end with "high" or "low" vowel sounds.

And the same process but different is repeated for regular words, since like many languages the way you say a letter or number's name is different from how you say it when it's in the middle of a word. "Exit 100A" is not the same thing as "Exit Hundred Aardvark" especially when the question is "what are the beginning and ending vowel sounds when spoken"

Reply:
This thing is called vowel harmony, and it's awfully complex. Usually translators solve it by rephrasing the sentence instead of writing long regexes. E.g. you can say `Forduljon ide: M4 autópálya` instead of `Forduljon az M4 Autópályára`.

https://en.wikipedia.org/wiki/Vowel_harmony#Hungarian

Also that's the recommendation of the Hungarian translators of fsf, best practices in Hungarian: https://fsfhu.gitlab.io/forditas-hogyan/#a-s-hez

Reply:
@infeeeee I'm aware, but we've done a really good job of making this sound natural in Hungarian per multiple conversations with multiple native speakers so I'm inclined to not just delete it all because of a few for loops

Reply:
1. Shouldn't the Hungarian TTS solve some of the issues you're trying to resolve manually?
2. Please put all these detailed comments into the corresponding code lines, so anyone else can later understand this "magic" without asking you again. Try to document any similar unclear places too. It is very important and helpful. After seeing/understanding the logic behind each line of your code, a better implementation can be suggested.

Reply:
> Also that's the recommendation of the Hungarian translators of fsf, best practices in Hungarian: https://fsfhu.gitlab.io/forditas-hogyan/#a-s-hez

Fwiw, those recommendations are for translators dealing with systems where natural translation isn't possible, so it lists some workarounds.

Reply:
@aronkvh OSMAnd uses the same methods for reading the street names, (to use the same example it says `M4 autópálya felé` ), so it's also good middle ground for those who don't want to deal with the complexity of agglutinative languages. 

Tested the current implementation, and it's working, I tried to find edge cases where it breaks, but couldn't find any, congrats!

Reply:
@biodranik unfortunately even Google doesn't handle grammar only pronunciation. So we don't need to write out "one hundred twenty three" at least but everything else will be read literally like a naive student with a really good accent would. Consider that TTS won't correct you if you write "she was a idiot" instead of "she was an idiot," this is the same situation just moreso.

Reply:
@infeeeee since when was OsmAnd announcing street names! Apparently I have to check it out again :P

Reply:
@biodranik comments added